src/System.Runtime/ref/System.Runtime.cs

```diff
@@ -2131,6 +2155,7 @@ public partial struct Single : System.IComparable, System.IComparable<float>, Sy
     }
     public readonly ref partial struct Span<T>
     {
+        private readonly int _dummy;
```

Comment:
I think .NET implementation contains an object. And even the fast implementation technically may contain an object reference.

We do not consider managed structs as unmanaged right now, but might in the future. 

I think Span/ReadOnlySpan should be considered object-containing

Reply:
> I think Span/ReadOnlySpan should be considered object-containing

I agree. It is not something that the tool can figure out from CoreLib metadata because of Span is treated as intrinsic by the runtime. Types like these will need exceptions. 

BTW: It is a bit odd that the primitive types like Int64 has 32-bit fields now. I won't be surprised if we find that it broke something somewhere and we need similar exceptions for them.

Reply:
On the other hand - even if we see all the fields we would consider fast span unmanaged. (since we do not see or understand the hidden byref)
And other ones - like `TypedReference` were always considered unmanaged.

I guess the object inside the slow spans would be the only reason to make the spans managed. - Just to conservatively match  the implementations.


---
In reply to: [161045136](https://github.com/dotnet/corefx/pull/26264#discussion_r161045136) [](ancestors = 161045136)

Reply:
Yeah - I had that concern about primitives as well. Maybe it would be safer if they are self-containing, like all their implementations?

---
In reply to: [161048209](https://github.com/dotnet/corefx/pull/26264#discussion_r161048209) [](ancestors = 161048209)

Reply:
I.E.  -   `int`contains an `int` dummy, `bool` contains a `bool`, and so on ...

---
In reply to: [161049185](https://github.com/dotnet/corefx/pull/26264#discussion_r161049185) [](ancestors = 161049185,161048209)

Reply:
> And other ones - like TypedReference were always considered unmanaged.

While it actually contains managed references because of it is treated as intrinsic by the runtime...

This highlights that this has number of false positives and false negatives.

Would it make sense to relax C# to let you take pointer from anything, and demote the current error for taking pointer of "managed type" to just a warning?

Reply:
Possibly. I do not know the exact rationale for not allowing pointers to managed structs. It was the case since v1.0

---
In reply to: [161052501](https://github.com/dotnet/corefx/pull/26264#discussion_r161052501) [](ancestors = 161052501)

Reply:
Latest update fixes primitive types to have matching field types and manually marks Span and ReadOnlySpan as having an object type.

Reply:
Are there any other special types we need to manually fix?

Reply:
I think this is all that is needed. 
The old-style ref-likes  like TypedReference should stay as they are. At least for compat reasons.

---
In reply to: [161073853](https://github.com/dotnet/corefx/pull/26264#discussion_r161073853) [](ancestors = 161073853)